### ISSUE TITLE
Proposal: smarter change detection

### ISSUE THREAD

**Issue by anjmao** [angular/angular#22646]
I created sad-hover component which is really sad at the moment. Here is stackblitz https://stackblitz.com/edit/angular-sad-hover?file=app%2Fapp.component.ts Open developer tools to see console output.

Problem:
1. I have isolated `sad-hover` component which uses OnPush.
2. Inside it I use (mouseleave) and (mouseover) and I don't want to trigger my whole app change detection to run.
3. This component is used in parent `sad-cat` component, and `sad-cat` is used on root app component.
3.  It turns out that zone is still grabbing these events and triggers parent component and root to update.

Solution:
Allow to have truly isolate component so everything that I make inside will stay inside, but I'm still be able to notify parent components about changes by using `@Output()`;
One solution could be to allow inject zone and say than I don't want to trigger any parent's component's CD.

```ts
class MyIsolatedCmp {
   constructor(zone: NgZone) {
      zone.SkipParent();
   }
}
```

**Comment by robert589** [comment]
why not u detach the component first? 
```
import { Component, ChangeDetectionStrategy, forwardRef, ChangeDetectorRef, ViewChild, ElementRef, NgZone } from '@angular/core';
import { FormsModule, ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

@Component({
  selector: 'my-app',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <sad-cat></sad-cat>
    <div *ngIf="isChangeDetectorTriggered()"></div>
  `
})
export class AppComponent  {
  isChangeDetectorTriggered() {
    console.log('root change detection :(');
    return true;
  }
}

@Component({
  selector: 'sad-cat',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <sad-hover></sad-hover>
    <div *ngIf="isChangeDetectorTriggered()"></div>
  `
})
export class SadCat {
  isChangeDetectorTriggered() {
    console.log('SadCat change detection :(');
    return true;
  }
}

@Component({
  selector: 'sad-hover',
  changeDetection: ChangeDetectionStrategy.OnPush,
  styles: [
    '.hover { background: #ddd}'
  ],
  template: `
    <div
    [class.hover]="hovered"
    (mouseleave)="onMouseleave()" 
    (mouseover)="onMouseover()">
      Hover me :(
    </div>
  `
})
export class SadHover {
  hovered = false;

  constructor(private _cdf : ChangeDetectorRef)   {
    this._cdf.detach();
  }
  onMouseover() {
    this.hovered = true;
    this._cdf.detectChanges();
  }

  onMouseleave() {
    this.hovered = false;
    this._cdf.detectChanges();
  }
}
```

**Comment by robert589** [comment]
Here is the solution, u have to detach the component using ChangeDetectionRef
```
import { Component, ChangeDetectionStrategy, forwardRef, ChangeDetectorRef, ViewChild, ElementRef, NgZone } from '@angular/core';
import { FormsModule, ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

@Component({
  selector: 'my-app',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <sad-cat></sad-cat>
    <div *ngIf="isChangeDetectorTriggered()"></div>
  `
})
export class AppComponent  {
    constructor(private _cdf : ChangeDetectorRef)   {
    this._cdf.detach();
  }
  isChangeDetectorTriggered() {
    console.log('root change detection :(');
    return true;
  }
}

@Component({
  selector: 'sad-cat',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <sad-hover></sad-hover>
    <div *ngIf="isChangeDetectorTriggered()"></div>
  `
})
export class SadCat {
  constructor(private _cdf : ChangeDetectorRef)   {
    this._cdf.detach();
  }
  isChangeDetectorTriggered() {
    console.log('SadCat change detection :(');
    return true;
  }
}

@Component({
  selector: 'sad-hover',
  changeDetection: ChangeDetectionStrategy.OnPush,
  styles: [
    '.hover { background: #ddd}'
  ],
  template: `
    <div
    [class.hover]="hovered"
    (mouseleave)="onMouseleave()" 
    (mouseover)="onMouseover()">
      Hover me :(
    </div>
  `
})
export class SadHover {
  hovered = false;

  constructor(private _cdf : ChangeDetectorRef)   {
    this._cdf.detach();
  }
  onMouseover() {
    this.hovered = true;
    this._cdf.detectChanges();
  }

  onMouseleave() {
    this.hovered = false;
    this._cdf.detectChanges();
  }
}
```

**Comment by anjmao** [comment]
@robert589 This is really bad solution. Why would you want to completely turn off change detection for other components? You then will need to manually manage all change detection for whose components.

**Comment by jotatoledo** [comment]
I dont have much experience with NgZone or zones.js, but I tried with this:

**Comment by varnastadeus** [comment]
I can confirm that  rxjs `fromEvent` also not triggers change detection.
```javascript
fromEvent(this.div.nativeElement, 'mouseover').subscribe(() => {
    this.onMouseover();
});
```

https://stackblitz.com/edit/angular-sad-hover-smbufy?file=app/app.component.ts

**Comment by sarunint** [comment]
You should read [this blog](https://blog.angularindepth.com/everything-you-need-to-know-about-change-detection-in-angular-8006c51d206f), it'll tell you how the change detection actually works.

**Comment by alexeagle** [comment]
Can anyone confirm/deny whether there's an actionable bug or feature request for Angular here?

**Comment by Toxicable** [comment]
I don't see a feature or bug.
The first demo is working as expected.
@anjmao When you trigger CD on a nested component CD must be run on the path down to the component. If you don't do this then you might easily violate one way data flow.
Using OnPush you can tell Angualr not to check a sibling sub tree but if there's only one path OnPush is not much help.

**Comment by anjmao** [comment]
@alexeagle It is not a bug. Example

```ts
@Component({
...
})
class MyCmp {
    constructor(private sharedService: SharedService) {}
    onHover() {
        this.sharedService.updateData(...);
    }
}
```
In this case CD should be triggered for parent because MyCmp is updating some shared state using services.

My proposal is about allowing to say: Look here is my component which is not updating any shared state and everything what will happen inside will not impact any parent component. As we see currently there are only workarounds to bind directly to nativeElement's event so CD is not triggered for parent.

**Comment by kemsky** [comment]
@Toxicable, the problem is that there is no easy way to skip change detection when listener is added in the component template i.e. in case you *know* that there will be no side effects. Obviously you can add listeners outside of zone using renderer api, manually detect changes and so on, but it is too complicated.

Simple example, assume we have a toolbar button Refresh, it can change its state (using js api, not CSS pseudo classes) hover/disabled/active/whatever and can trigger action - data modification. Only data modification (click event) should trigger change detection for the whole application, mouseover/mouseout events should not trigger global change detection in this case. This effects can be easily observed in a large application.

**Comment by mlc-mlapis** [comment]
@kemsky ... I think that all of us understand what it is the scenario but the question is if such a new API feature opens a door to something what could be really fast out of control when used massively in apps. I mean increasing number of reported issues which won't be issues at all but just misunderstandings or wrong using.

**Comment by anjmao** [comment]
@mlc-mlapis There will always be somebody who don't understand how things work. Worrying about reported issues is the negligible problem.

**Comment by LinBoLen** [comment]
1. detect change is not the fashion. time has passed, no another well known project stand with detect change. 
2. the growing solution or alternative solve binding afaik, @Bindable, reactivex, vdom, use object.defineproperty get set, @computed() @observer. maybe another i don't know.
3. the benefit of detectchange is that it can as a global schedule, and tick as async change.
4. the weak of detectchange is that it is the core of angular. so another library works with angular, must consider detect change.

finally, I wan to show my option that detect change is not so gooood and also not influence acceptable.

**Comment by mhagnumdw** [comment]
@varnastadeus 
> I can confirm that rxjs `fromEvent` also not triggers change detection.

I do not confirm that. For me `fromEvent` triggered the change detection.
rxjs: 6.2.2
